numpyro/contrib/funsor/infer_util.py

```diff
@@ -70,6 +71,28 @@ def config_fn(site):
     return infer_config(fn, config_fn)
 
 
+def compute_markov_factors(time_to_factors, markov_vars, sum_vars, prod_vars):
+    markov_factors = []
+    for time_var, log_factors in time_to_factors.items():
+        prev_vars = markov_vars[time_var]
+        prev_to_curr = {}
+        curr_vars = []
+        for k in prev_vars:
+            curr_var = "/".join(k.split("/")[1:])
+            curr_vars.append(curr_var)
+            prev_to_curr[k] = curr_var
+        prev_and_curr = prev_vars | frozenset(curr_vars)
+        with funsor.interpreter.interpretation(funsor.terms.lazy):
+            eliminate = (sum_vars | prod_vars) - prev_and_curr
```

Comment:
Time variables do not appear in `sum_vars` and `prod_vars` so I think this is fine.

Reply:
If you have a time-global variable that couples a bunch of time-local variables (like a global parameter in a state-space model's transition matrix), then eliminating that global variable couples all of the local variables and breaks the conditional independence that allows the use of `sequential_sum_product`.  I think you'll need to run `sequential_sum_product` before this step, rather than after.

Reply:
Hmm, I think I understand this wrong. Using `eliminate=frozenset()` passes `plate` and global discrete tests.

Reply:
Yes, when `eliminate == frozenset()` this works fine, but when that's not the case it may be incorrect to sum out variables in this order.

Reply:
It seems that the test for multiple plates at the same dimension does not pass due to this issue. I think I need to follow your strategy: using `sequential_sum_product` before this step. Did you suggested that we need to eliminate time dimensions in `x[t]|x[t-1]` and `y[t]|y[t-1]` separately?

Reply:
Yes, so I think we need to set eliminate=frozenset(). The tricky thing is when we have two plate statements at dim=-1, z[t]|x[t] and y[t]|z[t], we need to concatenate z and y, eliminate dimensions, then split the result. The current code will allocate two separate dimensions for two plates, which leads to wrong result. Do you have a solution for this?

Reply:
Because we don't eliminate any variable, I think we can use `sum(log_factors)` here for simplicity. However, we still need to have a mechanism to merge variables that allocate the same dimension. It is the job of `sequential_sum_product` I guess?

Reply:
> However, we still need to have a mechanism to merge variables that allocate the same dimension

Sorry, I'm not sure I understand the question.  What do you mean here? What is the difference between `z[t, i]` and `z[t]`, which I assume are supposed to be `.inputs` of a factor funsor?  In this function, all of the log-factors are funsors, so I don't see why you can't just sum the log-factors.

Reply:
Yes, we can take the sum, but assume z has addition plate D1 at dim=-1 and y has addition plate D2 at dim=-1, taking the sum will return a funsor with inputs time x prev x curr x D1 x D2. But I think the correct behavior is time x prev x curr x (D1 + D2).

Reply:
> but assume z has addition plate D1 at dim=-1 and y has addition plate D2 at dim=-1

Sorry, I still don't understand what you mean when you say that a funsor input `z` has an additional plate at dim=-1. Inputs are just inputs; all plates should be included in the log-factor funsors as additional inputs.  Can you give a concrete example set of log-factor funsors that exhibit this behavior?

Reply:
I see, I understand the situation you have in mind in `test_scan_enum_plates_same_dim`. I don't see why the fact that the plates share a dimension (dim=-1) should matter from the perspective of funsor, since everything is named and unordered.  However, I do understand now what you mean about stacking the plate dimensions (`D1 + D2`) rather than broadcasting to the cartesian product (`D1 x D2`).  

You can run tensor variable elimination (`funsor.sum_product.sum_product`) to eliminate those plates before applying `sequential_sum_product`, as you were doing before.  This would have the same effect as stacking the plates, but there are subtle restrictions on what can happen inside of them in the model for variable elimination to remain tractable - conservatively, any enumeration inside should be restricted to a single timestep (so the value of an enumerated sample site within one of the plates may not leave it or be used at the next timestep).  The situation in `test_scan_enum_plates_same_dim` is a special case of this where no new variables are enumerated within either plate.

Reply:
Thanks a lot, @eb8680!!! I just eliminate all dimensions that do not appear in markov sites (discrete sites that depend on the last step) and ... all test cases that I can think about passed. Could you take a look at the tests and suggest to me other scenarios that we need to tests?

Reply:
Looking like I am ready to port `hmm_enum` example now. :D thanks for your help!